src/main/java/org/spongepowered/common/data/manipulator/immutable/tileentity/ImmutableSpongeFurnaceData.java

```diff
@@ -0,0 +1,98 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.manipulator.immutable.tileentity;
+
+import com.google.common.collect.ComparisonChain;
+
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.MemoryDataContainer;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.immutable.tileentity.ImmutableFurnaceData;
+import org.spongepowered.api.data.manipulator.mutable.tileentity.FurnaceData;
+import org.spongepowered.api.data.value.immutable.ImmutableBoundedValue;
+import org.spongepowered.common.data.manipulator.immutable.common.AbstractImmutableData;
+import org.spongepowered.common.data.manipulator.mutable.tileentity.SpongeFurnaceData;
+import org.spongepowered.common.data.util.ComparatorUtil;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeBoundedValue;
+
+public class ImmutableSpongeFurnaceData extends AbstractImmutableData<ImmutableFurnaceData, FurnaceData> implements ImmutableFurnaceData {
+
+    private final int remainingBurnTime;
+    private final int remainingCookTime;
+
+    public ImmutableSpongeFurnaceData(int burnTime, int cookTime) {
+        super(ImmutableFurnaceData.class);
+        this.remainingBurnTime = burnTime;
+        this.remainingCookTime = cookTime;
+        this.registerGetters();
+    }
+
+    @Override
+    protected void registerGetters() {
+        this.registerFieldGetter(Keys.REMAINING_BURN_TIME, ImmutableSpongeFurnaceData.this::getRemainingBurnTime);
+        this.registerKeyValue(Keys.REMAINING_BURN_TIME, ImmutableSpongeFurnaceData.this::remainingBurnTime);
+
+        this.registerFieldGetter(Keys.REMAINING_COOK_TIME, ImmutableSpongeFurnaceData.this::getRemainingCookTime);
+        this.registerKeyValue(Keys.REMAINING_COOK_TIME, ImmutableSpongeFurnaceData.this::remainingBurnTime);
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> remainingBurnTime() {
+        return ImmutableSpongeBoundedValue.cachedOf(Keys.REMAINING_BURN_TIME, 0, this.remainingBurnTime, ComparatorUtil.intComparator(), 0, Integer.MAX_VALUE);
```

Comment:
Since the underlying value is an `Integer`, you shouldn't be creating a cached instance. There's a very large number of possible values, which will cause `ImmutableDataCachingUtil` to hold onto many different value instances, which will only be used once.

Instead, store an `ImmutableSpongeBoundedValue` in a field, and return in in this method.